File: rust/step1_read_print.rs
extern crate rustyline;
extern crate regex;
extern crate lazy_static;

use rustyline::error::ReadlineError;
use rustyline::Editor;

mod types;
mod reader;
mod printer;


struct Ast {
    tokens: String
}

fn read(input: String) -> Ast {
    let test = reader::read_str(input);
    println!("{:?}", test.tokens);
    Ast { tokens: String::from("hello") }
}

fn eval(ast: Ast) -> String {
    ast.tokens
}

fn print(exp: String) -> String {
    exp
}

fn rep(line: String) -> String {
    return print(eval(read(line)));
}

fn main() {
    let mut rl = Editor::<()>::new();
    if rl.load_history("history.txt").is_err() {
        println!("No previous history.");
    }

    loop {
        let readline = rl.readline("user> ");
        match readline {
            Ok(line) => {
                rl.add_history_entry(line.as_str());
                let exp = rep(line);
                println!("{}", exp);
            },
            Err(ReadlineError::Interrupted) => {
                println!("CTRL-C");
                break
            },
            Err(ReadlineError::Eof) => {
                println!("CTRL-D");
                break
            },
            Err(err) => {
                println!("Error: {:?}", err);
                break
            }
        }
        rl.save_history("history.txt").unwrap();
    }
}

File: es6c/step1_read_print.mjs

import rl from './node_readline.js'
import { read_str } from "./reader.mjs"
import { pr_str } from "./printer.mjs"

const { readline } = rl;

// read
const READ = str => read_str(str)

// eval
const EVAL = (ast, env) => ast

// print
const PRINT = exp => pr_str(exp)

// repl
const REP = str => PRINT(EVAL(READ(str)), {})

while (true) {
    let line = readline('user> ')
    if (line == null) break
    try {
        if (line) { console.log(REP(line)) }
    }
    catch (exc) {
        console.warn(`Error: ${exc}`)
    }
    
}